include/odp/api/spec/proto_stats.h

```diff
@@ -0,0 +1,151 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright(C) 2021 Marvell.
+ */
+
+/**
+ * @file
+ *
+ * ODP Proto Stats
+ */
+
+#ifndef ODP_API_SPEC_PROTO_STATS_H_
+#define ODP_API_SPEC_PROTO_STATS_H_
+#include <odp/visibility_begin.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @addtogroup odp_proto_stats
+ *  @{
+ */
+
+/**
+ * @def ODP_PROTO_STATS_INVALID
+ * Invalid proto stats handle
+ */
+
+/** ODP proto stats id */
+typedef enum odp_proto_stats_id_t {
+	/** None */
+	ODP_PROTO_STATS_ID_NONE,
+} odp_proto_stats_id_t;
+
+/** ODP proto stats mode */
+typedef enum odp_proto_stats_mode_t {
+	ODP_PROTO_STATS_MODE_MAX,
+} odp_proto_stats_mode_t;
+
+/** ODP proto stats params */
+typedef struct odp_proto_stats_param_t {
+	/** Mode of the proto stats object */
+	odp_proto_stats_mode_t mode;
+} odp_proto_stats_param_t;
+
+/**
+ * Proto stats capabilities
+ */
+typedef struct odp_proto_stats_capa_t {
+	/** Supported proto stat modes */
+	odp_bool_t modes[ODP_PROTO_STATS_MODE_MAX];
+} odp_proto_stats_capa_t;
+
+/**
+ * Initialize proto stats parameters
+ *
+ * Initialize an odp_proto_stats_param_t to its default values.
+ *
+ * @param param   Proto stats parameter pointer.
+ */
+void odp_proto_stats_init(odp_proto_stats_param_t *param);
+
+/**
+ * Get proto stats capability
+ *
+ * Get proto stats capability of supported modes.
+ *
+ * @param capa Pointer where capabilities are updated.
+ */
+void odp_proto_stats_capa(odp_proto_stats_capa_t *capa);
+
+/**
+ * Create a proto stats object
+ *
+ * Create a proto stats object with given name and mode.
+ * Semantics of each mode is associated with that mode.
+ *
+ * @param name  Object name
+ * @param param Proto stats parameters
+ *
+ * @return Proto stats object handle
+ */
+odp_proto_stats_t odp_proto_stats_create(const char *name, const odp_proto_stats_param_t *param);
+
+/**
+ * Lookup a proto stats object by name
+ *
+ * Lookup an already created proto stats object by name.
+ *
+ * @param name Proto stats object name
+ *
+ * @return Proto stats object handle
+ */
+odp_proto_stats_t odp_proto_stats_lookup(const char *name);
+
+/**
+ * Destroy a proto stats object
+ *
+ * Destroy a proto stats object already created.
+ *
+ * @param stat Proto stats handle
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_proto_stats_destroy(odp_proto_stats_t stat);
```

Comment:
When is it safe to call the destroy? Any time? When no packets that refer to the proto stat are no longer in flight (but how would one know it?)?

Reply:
Our thought is to preallocate stats objects and use it. Hence there won't be any segfaults and application can free destroy odp_proto_stats_t anytime. But in order to get the correct updates, they need to stop pktio which ensures all the packets are flushed out.

Reply:
If I create a proto stats object, associate a packet with it, then delete the proto stats object and create a new one, it is possible that the packet that still refers to the old, now stale, proto stats object might end up updating the newly added proto stats object and screw up its statistics? I would suppose preventing this would require some extra logic in the ODP implementation.

Reply:
Yes. If we have to handle synchronization in ODP implementation, it would be requiring costly  logic either in implementation and impacts fast path perf.

Reply:
Such synchronization would seem to be needed with the current API draft since the API seems to allow calling the destroy at any time. If this is not ok, then you probably want specify in the API when the proto stats object may be destroyed.

Reply:
Ack, I'll update the documentation.